src/structures/amd_sev.rs

```diff
@@ -0,0 +1,129 @@
+use core::sync::atomic::AtomicBool;
+use core::sync::atomic::Ordering::SeqCst;
+
+use bit_field::BitField;
+
+use crate::registers::model_specific::Msr;
+use crate::structures::paging;
+use crate::structures::paging::PageTableFlags;
+
+const MSR_AMD_SEV: Msr = Msr::new(0xc0010131);
+
+const SEV_INITIALIZED: AtomicBool = AtomicBool::new(false);
+
+
+/// Retrieves the AMD SEV status, after calling `init`
+pub fn sev_state<'a>() -> Option<&'a SevState> {
+	AMD_SEV_STATUS.as_ref()
+}
+
+/// Initializes AMD Secure Encrypted Virtualization, if enabled.
+///
+/// It is required to call this function in an SEV enabled guest, as it sets-up a dynamic page table
+/// flag for memory encryption. If this is not set, page table operations may cause panics due to
+/// invalid canonical addresses.
+pub fn init<'a>() -> Option<&'a SevState> {
+	if SEV_INITIALIZED.fetch_or(true, SeqCst) {
+		return sev_state();
+	}
+
+	// https://github.com/torvalds/linux/blob/900241a5cc15e6e0709a012051cc72d224cd6a6e/arch/x86/mm/mem_encrypt_identity.c#L566
+	// Check for SME support
+	let (cpuid_max, _) = unsafe { core::arch::x86_64::__get_cpuid_max(0x8000_0000) };
+	if cpuid_max < 0x8000_001F {
+		return None;
+	}
+
+	// Check if SME is available on the current CPU then read the C-Bit position and define the mask
+	let sme_features = unsafe { core::arch::x86_64::__cpuid(0x8000_001F) };
+	let sme_available = (sme_features.eax & 0x3) != 0; // either SME or SEV bit is set
+	if !sme_available {
+		return None;
+	}
```

Comment:
On SEV-ES/SEV-SNP, using the CPUID instruction causes a #VC exception, and I'd like to prefer to avoid this if possible. Handling a #VC exception is non-trivial, and some people might choose not to do this.

Instead of determining the C-bit ourselves, how about we ask the user to provide the bit position instead? That'd probably also mean that we'd have to mark the init function as unsafe, but realistically, writing a #VC handler is going to involve even more unsafe code. We could also provide both.

Note that this approach would also work for Intel TDX. AFAICT on TDX, there's no easy way to determine the bit position of the S-bit (at least none that always works), but the bit position is given in the RBX register at VM launch, so the user already has this information.

Reply:
This is a good idea yes, since OSes will need their SEV initialization procedures anyway.